ДОМАШНЯЯ РАБОТА № 1 

Ходыка Анны

In [102]:
!pip install numpy

In [6]:
import numpy as np

Задача 1. Вкладчик положил 50000 условных единиц на три разных счета в три разных банка. По первому счету выплаты составят 5% годовых, по второму – 7% годовых и по третьему 6% годовых. Известно, что через год вкладчик получил по процентам сумму в 2250 у.е. с первого и второго банка и сумму в 1400 у.е. с первого и третьего банка. Сколько условных единиц он положил на каждый счет изначально?

In [20]:
matrix = np.array([[1,1,1],[0.05,0.07,0], [0.05,0,0.06]], float)
matrix

array([[1.  , 1.  , 1.  ],
       [0.05, 0.07, 0.  ],
       [0.05, 0.  , 0.06]])

In [21]:
arr=np.array([50000,2250,1400], float)
arr

array([50000.,  2250.,  1400.])

In [24]:
answer = np.linalg.solve(matrix, arr)
answer

array([10000., 25000., 15000.])

Вкладчик положил:
10000 условных единиц на первый счет,
25000 условных единиц на второй счет,
15000 условных единиц на третий счет.
Проверка правильности ответов:
10000+25000+15000=50000
10000*5%+25000*7% =500+1750=2250 
10000*5%+15000*6% = 500+900=1400

In [31]:
assert np.dot(matrix, answer).all() == arr.all()

Задача 2. На складе хранятся мобильные телефоны iPhone 6, iPhone 11, iPhone 12 моделей. Всего 1328 экземпляра. Моделей iPhone 6 на 120 меньше, чем моделей iPhone 11, и на 100 больше, чем моделей iPhone 12. Сколько моделей каждого вида на складе?

In [33]:
phones_matrix = np.array([[1,1,1],[-1,1,0],[1,0,-1]], float)
phones_matrix

array([[ 1.,  1.,  1.],
       [-1.,  1.,  0.],
       [ 1.,  0., -1.]])

In [34]:
phones_b=np.array([1328,120,100], float)
phones_b

array([1328.,  120.,  100.])

In [35]:
phones_answer = np.linalg.solve(phones_matrix, phones_b)
phones_answer 

array([436., 556., 336.])

На складе хранятся мобильные телефоны iPhone 6 - 436 штук, iPhone 11 - 556 штук, iPhone 12 моделей - 336 штук.
Проверка:

In [36]:
assert np.dot(phones_matrix, phones_answer).all() == phones_b.all()

Задача 3. Составьте каноническое уравнение эллипсоида, проходящего через три точки A,B,C

In [120]:
elipsoid_matrix = np.array([[3,0,3], [6,1/4,0],[1,1/3,1]])
elipsoid_b=np.array([1,1,1])
elipsoid_answer = np.linalg.solve(elipsoid_matrix, elipsoid_b)
elipsoid_coef = (1/elipsoid_answer)**0.5
elipsoid_coef

array([3.46410162, 0.70710678, 2.        ])

In [126]:
# проверка
x = 3**0.5
y=0
z=3**0.5
assert round(x**2/(3.46410162**2)+y**2/(0.70710678**2)+z**2/(2**2), 2) ==1
x = 6**0.5
y=1/2
z=0
assert round(x**2/(3.46410162**2)+y**2/(0.70710678**2)+z**2/(2**2), 2) ==1
x = 1
y=1/(3**0.5)
z=1
assert round(x**2/(3.46410162**2)+y**2/(0.70710678**2)+z**2/(2**2), 2) ==1

Задача 4. Составьте уравнение параболы проходящей через заданные три точки (1,12), (3,54), (-1,2). Необходимо найти коэффициенты уравнения a,b,c
y=a⋅x^2+b⋅x+c

Подставляю координаты 3 точек в уравнение, получаю СЛАУ с неизвестными a,b,c:

In [38]:
parabola_matrix = np.array([[1,1,1],[9,3,1],[1,-1,1]], float)
parabola_matrix

array([[ 1.,  1.,  1.],
       [ 9.,  3.,  1.],
       [ 1., -1.,  1.]])

In [40]:
parabola_b = np.array([12,54,2], float)
parabola_b

array([12., 54.,  2.])

In [42]:
parabola_answer = np.linalg.solve(parabola_matrix, parabola_b)
parabola_answer

array([4., 5., 3.])

Уравнение 4x^2+5x+3=y

Проверка:

In [43]:
assert np.dot(parabola_matrix, parabola_answer).all() == parabola_b.all()

Задача 5. Составьте функцию get_polynom, которая по набору точек вида будет находить коэффициенты многочлена. Таким образом функция принимает список кортежей с координатами и возвращает набор коэффициентов

In [128]:
import numpy as np


def get_polynom(coords):
    n = len(coords)-1
    
    xy=np.array(coords)
    # присваеваем b - вектор из y
    b=xy[:, 1]
    x=xy[:, 0]
    x= x[:, np.newaxis]
    # формируем а - матрица из x в степени от 0 до n
    a=x**0
    for i in range(1,n+1):
        a= np.concatenate((a, x**i),axis=1)
    
    return np.linalg.solve(a, b)

In [129]:
get_polynom([(1,4), (0,1), (10, 121)])

array([1., 2., 1.])

ВТОРОЙ СПОСОБ: формирование a и b с помощью циклов for

In [100]:
import numpy as np


def get_polynom(coords):
    n = len(coords)-1
    list_for_a = []
    list_for_b = []
    
    for x, y in coords:
        list_for_b.append(y)
        x_list = []
        for i in range(n+1):
            x_list.append(x**i)
        list_for_a. append(x_list)    
    
    a = np.array(list_for_a)
    b = np.array(list_for_b)
    
    return np.linalg.solve(a, b)

In [131]:
get_polynom([(1,4), (0,1), (10, 121)])

array([1., 2., 1.])